### PR TITLE
Ruby: support conan v2 and expose more configuration options for native extension gems

### PR DESCRIPTION
**Pull request by jmarrec**
Specify library name and version:  **ruby/3.1.0**

Ruby: support conan v2 and expose more configuration options for native extension gems

Reboots:

* #12456
* #12208

Both PRs were opened more than a year ago and got stale and closed.

Changelog (old commit messages):

* conan v2 support (work done by myself + a merge of @SpaceIm 's branch from PR #12208 to reconcile both)

* Extend testing to require one of the native extensions
```
post_package(): WARN: [MISSING SYSTEM LIBS (KB-H043)] Library './lib/ruby/3.1.0/x86_64-linux/pty.so'
 links to system library 'util' but it is not in cpp_info.system_libs.
```

* Use `--with-opt-dir` instead of `--with-xx-dir` that ruby isn't respecting.

conanio/gccXX (eg 10) removed the libxxx-dev (eg libgmp-dev) from the image. This made me realize that the conan deps weren't being picked up by the build. The openssl was, because --with-openssl-dir is explicitly used in ruby config.

So here, we rely on --with-opt-dir.

cf: https://bugs.ruby-lang.org/issues/19014#change-99241 (--with-gmp-dir was added on September 14, 2022, wrote this in Sep 22, 2022)

* Pass both --with-opt-dir and --with-xxx-dir

https://gist.github.com/mrkn/6647630

---

- [x] I've read the [contributing guidelines](https://github.com/conan-io/conan-center-index/blob/master/CONTRIBUTING.md).
- [x] I've used a [recent](https://github.com/conan-io/conan/releases/latest) Conan client version close to the [currently deployed](https://github.com/conan-io/conan-center-index/blob/master/.c3i/config_v1.yml#L6). 
     - **Tested 1.60.1 and 2.0.7 locally**
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/ruby/all/test_package/conanfile.py [hunk 1]
```diff
@@ -1,21 +1,55 @@
...
+        # when --static-linked-ext is used, ruby defines EXTSTATIC as 1
+        # But when ruby itself is static there's nothing, so:
+        # We define RUBY_STATIC_RUBY when ruby itself is static
+        # We define RUBY_STATIC_LINKED_EXT when the ruby extensions are static (same as EXTSTATIC but clearer)
+        defs = {}
+        tested_options = self.dependencies[self.tested_reference_str].options
+        if not tested_options.shared:
+            defs["RUBY_STATIC_RUBY"] = 1
+            if tested_options.with_static_linked_ext:
+                defs["RUBY_STATIC_LINKED_EXT"] = 1
```
Comment (valgur) [inline recipes/ruby/all/test_package/conanfile.py:31]: Shouldn't these be added to `self.cpp_info.defines` in `package_info()` instead?

File: recipes/ruby/all/conanfile.py [hunk 2]
```diff
@@ -54,39 +60,70 @@ def _windows_system_libs(self):
...
             return "-O2sy-"
 
-    def requirements(self):
-        self.requires("zlib/1.2.12")
-        self.requires("gmp/6.1.2")
-        if self.options.with_openssl:
-            self.requires("openssl/1.1.1o")
+    def export_sources(self):
+        for p in self.conan_data.get("patches", {}).get(self.version, []):
+            copy(self, p["patch_file"], self.recipe_folder, self.export_sources_folder)
```
Comment (valgur) [inline recipes/ruby/all/conanfile.py:70]:
```suggestion
        export_conandata_patches(self)
```

File: recipes/ruby/all/conanfile.py [hunk 3]
```diff
@@ -54,39 +60,70 @@ def _windows_system_libs(self):
...
+        if self.options.with_openssl:
+            self.requires("openssl/1.1.1q")
+
+        if self.options.with_libyaml:
+            self.requires("libyaml/0.2.5")
+
+        if self.options.with_libffi:
+            self.requires("libffi/3.4.2")
+
+        if self.options.with_readline:
```
Comment (valgur) [inline recipes/ruby/all/conanfile.py:106]:
```suggestion
        if self.options.get_safe("with_readline"):
```

File: recipes/ruby/all/conanfile.py [hunk 4]
```diff
@@ -1,22 +1,18 @@
...
-    from conan.tools.build.cross_building import cross_building
-
-from conan.tools.files import apply_conandata_patches
-from conan.tools.gnu import Autotools, AutotoolsDeps, AutotoolsToolchain
-from conan.tools.microsoft import msvc_runtime_flag, is_msvc
-from conans import tools
-from conans.errors import ConanInvalidConfiguration
-
-required_conan_version = ">=1.43.0"
+required_conan_version = ">=1.51.3"
```
Comment (valgur) [inline recipes/ruby/all/conanfile.py:15]:
```suggestion
required_conan_version = ">=1.53"
```
For `self.settings.rm_safe()` support.

File: recipes/ruby/all/conanfile.py [hunk 5]
```diff
@@ -122,75 +178,97 @@ def generate(self):
...
+        if self.options.with_gmp:
+            self.cpp_info.requires.append("gmp::gmp")
+        if self.options.with_openssl:
+            self.cpp_info.requires.append("openssl::openssl")
+        if self.options.with_libyaml:
+            self.cpp_info.requires.append("libyaml::libyaml")
+        if self.options.with_libffi:
+            self.cpp_info.requires.append("libffi::libffi")
+        if self.options.with_readline:
+            self.cpp_info.requires.append("readline::readline")
```
Comment (valgur) [inline recipes/ruby/all/conanfile.py:270]:
```suggestion
```
Can be dropped, most likely.

File: recipes/ruby/all/conanfile.py [hunk 6]
```diff
@@ -122,75 +178,97 @@ def generate(self):
...
+        if is_apple_os(self):
+            self.cpp_info.frameworks = ["CoreFoundation"]
 
-        self.cpp_info.filenames["cmake_find_package"] = "Ruby"
-        self.cpp_info.filenames["cmake_find_package_multi"] = "Ruby"
+        self.cpp_info.set_property("cmake_find_mode", "both")
         self.cpp_info.set_property("cmake_file_name", "Ruby")
+        self.cpp_info.set_property("cmake_target_name", "Ruby::Ruby")
+        self.cpp_info.set_property("pkg_config_name", "ruby")
+        self.cpp_info.set_property("pkg_config_aliases", [f"ruby-{version.major}.{version.minor}"])
```
Comment (valgur) [inline recipes/ruby/all/conanfile.py:255]: I would move it to the top of `package_info()` as it's the most important bit of info to the package consumers, really, and it's kind of the conventional location in other recipes.

File: recipes/ruby/all/conanfile.py [hunk 7]
```diff
@@ -122,75 +178,97 @@ def generate(self):
...
+        # install the enc/*.a / ext/*.a libraries
+        if not self.options.shared and self.options.with_static_linked_ext:
+            for dirname in ['ext', 'enc']:
+                dst = os.path.join('lib', dirname)
+                copy(self, '*.a', src=dirname, dst=os.path.join(self.package_folder, dst), keep_path=True)
+                copy(self, '*.lib', src=dirname, dst=os.path.join(self.package_folder, dst), keep_path=True)
 
     def package_info(self):
         binpath = os.path.join(self.package_folder, "bin")
         self.output.info(f"Adding to PATH: {binpath}")
```
Comment (valgur) [inline recipes/ruby/all/conanfile.py:225]:
```suggestion
```
These logging statements are no longer recommended as they get logged even every time the package is used transitively.
Also, move `self.env_info.PATH.append(binpath)` under the `# TODO: to remove in conan v2` block, as it's no longer required in Conan v2.

File: recipes/ruby/all/test_package/conanfile.py (hunk 1 above)
Comment (jmarrec) [reply recipes/ruby/all/test_package/conanfile.py:34]: I honestly have no idea... you tell me! What I'm trying to do is to pass it to the CMakeLists.txt so they have defined in the executable: https://github.com/jmarrec/conan-center-index/blob/e5c6c958b0b90067ec7d93a468405095f4eff8be/recipes/ruby/all/test_package/CMakeLists.txt#L14-L20

And then: https://github.com/jmarrec/conan-center-index/blob/e5c6c958b0b90067ec7d93a468405095f4eff8be/recipes/ruby/all/test_package/test_package.cpp#L17-L27

**Comment by jmarrec** [comment]
@valgur Thanks for the review. I have address most of your comments. I'm kinda lost about to do with the test_package_v1, could you help please? I don't understand the error, and all of the test_v1_package folders from other recipes I've checked also have the outdated imports...

Thanks!

**Comment by jmarrec** [comment]
Oh, the pasted logs above were truncated. The error isn't the warnings about deprecated imports but:

```
ERROR: ruby/3.1.0 (test package): Error in build() method, line 30
	tested_options = self.dependencies[self.tested_reference_str].options
	AttributeError: 'NoneType' object has no attribute 'dependencies'
```

Still don't get the error though. Also, why is the v1 pipeline using this test_package folder and not the test_v1_package?

**Comment by jmarrec** [comment]
I don't get why the v2 pipeline is failing now when building shared.

But I can repro on conanio/gcc-11-ubuntu16.04. Opening `/home/conan/.conan2/p/t/ruby6bbd8583eb14e/b/build-release/config.log`

```
configure:5933: checking whether we are cross compiling
configure:5941: gcc -o conftest  -m64 -fPIC -O3  -I/home/conan/.conan2/p/readl9a10602e4f06c/p/include -I/home/conan/.conan2/p/gmpe88a5af46fc68/p/include -I/home/conan/.conan2/p/termc9b1a993805ca2/p/include -I/home/conan/.conan2/p/libff1634e8614167c/p/include -I/home/conan/.conan2/p/libya9cdeb62a42256/p/include -I/home/conan/.conan2/p/opens6b55f0181ad1f/p/include -I/home/conan/.conan2/p/zlib2bd9a64202691/p/include -DTERMCAP_SHARED -DNDEBUG  -L/home/conan/.conan2/p/readl9a10602e4f06c/p/lib -L/home/conan/.conan2/p/gmpe88a5af46fc68/p/lib -L/home/conan/.conan2/p/termc9b1a993805ca2/p/lib -L/home/conan/.conan2/p/libff1634e8614167c/p/lib -L/home/conan/.conan2/p/libya9cdeb62a42256/p/lib -L/home/conan/.conan2/p/opens6b55f0181ad1f/p/lib -L/home/conan/.conan2/p/zlib2bd9a64202691/p/lib -m64 conftest.c  -lhistory -lreadline -lgmpxx -lgmp -ltermcap -lffi -lyaml -lssl -lcrypto -lz -lm -ldl -lpthread -lrt >&5
configure:5945: $? = 0
configure:5952: ./conftest
./conftest: error while loading shared libraries: libhistory.so.8: cannot open shared object file: No such file or directory
configure:5956: $? = 127
configure:5963: error: in `/home/conan/.conan2/p/t/ruby6bbd8583eb14e/b/build-release':
configure:5965: error: cannot run C compiled programs.
If you meant to cross compile, use `--host'.
See `config.log' for more details
```

Yet, `-L/home/conan/.conan2/p/readl9a10602e4f06c/p/lib` is passed, and it does have it?

```
$ ls /home/conan/.conan2/p/readl9a10602e4f06c/p/lib
libhistory.so  libhistory.so.8  libhistory.so.8.1  libreadline.so  libreadline.so.8  libreadline.so.8.1
```

**Comment by jmarrec** [comment]
Same problem on master actually... with conan 1 (.62)

```
(conan1) conan@b07fc5063ca6:~/conan-center-index/recipes/ruby/all$ conan --version
Conan version 1.62.0
(conan1) conan@b07fc5063ca6:~/conan-center-index/recipes/ruby/all$ conan create --build missing -o '*/*:shared=True' . ruby/3.1.0@

[...]
checking whether we are cross compiling... configure: error: in `/home/conan/.conan/data/ruby/3.1.0/_/_/build/3b2a764fe6df91ad358a5488f30968aeb89a0099':
configure: error: cannot run C compiled programs.
If you meant to cross compile, use `--host'.
See `config.log' for more details
ruby/3.1.0:
ruby/3.1.0: ERROR: Package '3b2a764fe6df91ad358a5488f30968aeb89a0099' build failed
ruby/3.1.0: WARN: Build folder /home/conan/.conan/data/ruby/3.1.0/_/_/build/3b2a764fe6df91ad358a5488f30968aeb89a0099
ERROR: ruby/3.1.0: Error in build() method, line 139
	at.configure(build_script_folder=build_script_folder)
	ConanException: Error 1 while executing "/home/conan/.conan/data/ruby/3.1.0/_/_/build/3b2a764fe6df91ad358a5488f30968aeb89a0099/source_subfolder/configure" '--enable-shared' '--disable-static' '--prefix=/' '--bindir=${prefix}/bin' '--sbindir=${prefix}/bin' '--libdir=${prefix}/lib' '--includedir=${prefix}/include' '--oldincludedir=${prefix}/include' '--disable-install-doc'
```

config.log

```
configure:5889: $? = 0
configure:5911: result:
configure:5933: checking whether we are cross compiling
configure:5941: gcc -o conftest  -m64 -fPIC -O3  -I/home/conan/.conan/data/openssl/1.1.1o/_/_/package/1748639999ed79b998e4fe4a6d292ed8e874736a/include -I/home/conan/.conan/data/gmp/6.1.2/_/_/package/9572c3b88ec662d3a69e083d2310b7be12936bbf/include -I/home/conan/.conan/data/zlib/1.2.12/_/_/package/1748639999ed79b998e4fe4a6d292ed8e874736a/include -DNDEBUG  -L/home/conan/.conan2/p/readl9a10602e4f06c/p/lib -L/home/conan/.conan2/p/gmpe88a5af46fc68/p/lib -L/home/conan/.conan2/p/termc9b1a993805ca2/p/lib -L/home/conan/.conan2/p/libff1634e8614167c/p/lib -L/home/conan/.conan2/p/libya9cdeb62a42256/p/lib -L/home/conan/.conan2/p/opens6b55f0181ad1f/p/lib -L/home/conan/.conan2/p/zlib2bd9a64202691/p/lib -m64 -L/home/conan/.conan/data/openssl/1.1.1o/_/_/package/1748639999ed79b998e4fe4a6d292ed8e874736a/lib -L/home/conan/.conan/data/gmp/6.1.2/_/_/package/9572c3b88ec662d3a69e083d2310b7be12936bbf/lib -L/home/conan/.conan/data/zlib/1.2.12/_/_/package/1748639999ed79b998e4fe4a6d292ed8e874736a/lib -m64 conftest.c  -lssl -lcrypto -lgmpxx -lgmp -lz -ldl -lpthread -lrt -lm >&5
configure:5945: $? = 0
configure:5952: ./conftest
./conftest: error while loading shared libraries: libssl.so.1.1: cannot open shared object file: No such file or directory
configure:5956: $? = 127
configure:5963: error: in `/home/conan/.conan/data/ruby/3.1.0/_/_/build/3b2a764fe6df91ad358a5488f30968aeb89a0099':
configure:5965: error: cannot run C compiled programs.
If you meant to cross compile, use `--host'.
See `config.log' for more details
```

Can I get some help pretty please @conan-io/barbarians ?

**Comment by jmarrec** [comment]
Because it links with `-L/path/to/pkgname/lib -l<pkgname>`, the link info is just the soname (libpkgname.so). So when executing `conftest`, it can't find it, because that's not a system library. If I pass `LD_LIBRARY_PATH` it works.

```
LD_LIBRARY_PATH="/home/conan/.conan2/p/readl9a10602e4f06c/p/lib:/home/conan/.conan2/p/gmpe88a5af46fc68/p/lib:/home/conan/.conan2/p/termc9b1a993805ca2/p/lib:/home/conan/.conan2/p/libff1634e8614167c/p/lib:/home/conan/.conan2/p/libya9cdeb62a42256/p/lib:/home/conan/.conan2/p/opens6b55f0181ad1f/p/lib:/home/conan/.conan2/p/zlib2bd9a64202691/p/lib:/home/conan/.conan/data/openssl/1.1.1o/_/_/package/1748639999ed79b998e4fe4a6d292ed8e874736a/lib:/home/conan/.conan/data/gmp/6.1.2/_/_/package/9572c3b88ec662d3a69e083d2310b7be12936bbf/lib:/home/conan/.conan/data/zlib/1.2.12/_/_/package/1748639999ed79b998e4fe4a6d292ed8e874736a/lib" ./conftest
```

**Comment by valgur** [comment]
Try replacing the `AutotoolsDeps` with
```python
    def build_requirements(self):
        if not self.conf.get("tools.gnu:pkg_config", check_type=str):
            self.tool_requires("pkgconf/2.1.0")

    def generate(self):
        venv = VirtualBuildEnv(self)
        venv.generate()

        deps = PkgConfigDeps(self)
        deps.generate()
```

Shared build on Linux fails with `/lib/x86_64-linux-gnu/libtermcap.so: invalid ELF header` for me otherwise.

`pkg-config` is a somewhat preferable way of adding dependencies to an autotools project. AutotoolsDeps is a rather brute-force solution that blindly includes and links against everything.

**Comment by jmarrec** [comment]
I'll try that. Otherwise I was able to make it work with 

 I can fix it via 

```python
    def generate(self):
        td = AutotoolsDeps(self)
        td.environment.append(
            "LD_LIBRARY_PATH", [x for dep in td.ordered_deps for x in dep.cpp_info.libdirs], separator=":"
        )
```

but I can't find any precedence for this so I'm guessing this is hacky at best.

File: recipes/ruby/all/test_package/test_package.cpp [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+#endif
+
+int main(int argc, char* argv[]) {
+  ruby_sysinit(&argc, &argv);
+  RUBY_INIT_STACK;
+  ruby_init();
+
+  rb_eval_string("puts 'Hello, ruby!'");
+
+#ifdef RUBY_STATIC_RUBY
```
[the remark this answers is not in the document]
Comment (jmarrec) [reply recipes/ruby/all/test_package/test_package.cpp:17]: That's just print statements as an info/debug message. The bit of interest is below, with how the extension gem bigdecimal is loaded.

**Comment by jmarrec** [comment]
V2 pipeline; I’m having a (new to me) build failure on CI (macos): `/bin/sh: line 0: exec: autoconf: not found`
Recipe uses Autotools, PkgConfigDeps, AutotoolsToolchain , ins’t autoconf supposed to be there?

Do I need to explicitly add it as a build requirement?

```
def build_requirements(self):
        if self.settings.os != "Windows":
            self.tool_requires("autoconf/2.71")
```

V1: awesome, an internal compiler error.

**Comment by jmarrec** [comment]
Can't reproduce the ICE on either `conanio/gcc5-ubuntu16.04:1.60.0` nor `conanio/gcc9-ubuntu16.04:1.60.0` ...

**Comment by jmarrec** [comment]
I can reproduce the ICE on gcc 9 actually... Except when I go to the directory, and make again, this time it works fine...

```
make[1]: Entering directory '/home/conan/.conan/data/ruby/3.1.0/_/_/package/efe260f7679565dd5c8cb1f99b9f0ad6795646d3/home/conan/.conan/data/ruby/3.1.0/_/_/package/efe260f7679565dd5c8cb1f99b9f0ad6795646d3/lib/ruby/gems/3.1.0/gems/rbs-2.0.0/ext/rbs_extension'
compiling constants.c
compiling lexer.c
compiling lexstate.c
compiling location.c
compiling main.c
compiling parser.c
compiling parserstate.c
compiling ruby_objs.c
compiling unescape.c
during GIMPLE pass: unswitch
parserstate.c: In function 'parser_typevar_member':
parserstate.c:73:6: internal compiler error: Segmentation fault
   73 | bool parser_typevar_member(parserstate *state, ID id) {
      |      ^~~~~~~~~~~~~~~~~~~~~
mmap: Permission denied
Please submit a full bug report,
with preprocessed source if appropriate.
See <https://gcc.gnu.org/bugs/> for instructions.
Makefile:245: recipe for target 'parserstate.o' failed
make[1]: *** [parserstate.o] Error 1
make[1]: *** Waiting for unfinished jobs....
make[1]: Leaving directory '/home/conan/.conan/data/ruby/3.1.0/_/_/package/efe260f7679565dd5c8cb1f99b9f0ad6795646d3/home/conan/.conan/data/ruby/3.1.0/_/_/package/efe260f7679565dd5c8cb1f99b9f0ad6795646d3/lib/ruby/gems/3.1.0/gems/rbs-2.0.0/ext/rbs_extension'

make failed, exit code 2
```

```
conan@7f87163a9487:~/.conan/data/ruby/3.1.0/_/_/package/efe260f7679565dd5c8cb1f99b9f0ad6795646d3/home/conan/.conan/data/ruby/3.1.0/_/_/package/efe260f7679565dd5c8cb1f99b9f0ad6795646d3/lib/ruby/gems/3.1.0/gems/rbs-2.0.0$ cd /home/conan/.conan/data/ruby/3.1.0/_/_/package/efe260f7679565dd5c8cb1f99b9f0ad6795646d3/home/conan/.conan/data/ruby/3.1.0/_/_/package/efe260f7679565dd5c8cb1f99b9f0ad6795646d3/lib/ruby/gems/3.1.0/gems/rbs-2.0.0/ext/rbs_extension
conan@7f87163a9487:~/.conan/data/ruby/3.1.0/_/_/package/efe260f7679565dd5c8cb1f99b9f0ad6795646d3/home/conan/.conan/data/ruby/3.1.0/_/_/package/efe260f7679565dd5c8cb1f99b9f0ad6795646d3/lib/ruby/gems/3.1.0/gems/rbs-2.0.0/ext/rbs_extension$ make
compiling lexer.c
compiling lexstate.c
compiling location.c
compiling main.c
compiling parser.c
compiling ruby_objs.c
compiling unescape.c
linking shared-object rbs_extension.so
```

**Comment by jmarrec** [comment]
This is very frustrating. I worked around the ICE issue by bumping the "rbs" and "debug"  bundled gems uses the versions listed on the ruby_3_1 branch https://github.com/ruby/ruby/blob/ruby_3_1/gems/bundled_gems#L14-L16 ([permalink](https://github.com/ruby/ruby/blob/0a70dde90ab2b52191f4782d0ec7d0a99ab1e845/gems/bundled_gems#L14-L16)).

But there are still issues I really don't understand. Seems like the cross building from x86_64 mac to armv8 is failing.

```
ruby/3.1.0: Package 'be0af55287e289adf24870a7f45a92f7dffcc5a5' built
ruby/3.1.0: Build folder /Users/jenkins/w/prod-v1/bsr/74985/edbca/.conan/data/ruby/3.1.0/_/_/build/be0af55287e289adf24870a7f45a92f7dffcc5a5/build-release
ruby/3.1.0: Generated conaninfo.txt
ruby/3.1.0: Generated conanbuildinfo.txt
ruby/3.1.0: Generating the package
ruby/3.1.0: Package folder /Users/jenkins/w/prod-v1/bsr/74985/edbca/.conan/data/ruby/3.1.0/_/_/package/be0af55287e289adf24870a7f45a92f7dffcc5a5
ruby/3.1.0: Calling package()
ruby/3.1.0: Copied 1 file: COPYING
ruby/3.1.0: Copied 1 file: BSDL

----Running------
> . "/Users/jenkins/w/prod-v1/bsr/74985/edbca/.conan/data/ruby/3.1.0/_/_/build/be0af55287e289adf24870a7f45a92f7dffcc5a5/build-release/conan/conanbuild.sh" && make install-local -j8
-----------------
/Users/jenkins/w/prod-v1/bsr/74985/edbca/.conan/data/ruby/3.1.0/_/_/build/be0af55287e289adf24870a7f45a92f7dffcc5a5/src/revision.h unchanged
/usr/bin/ruby --disable=gems -I/Users/jenkins/w/prod-v1/bsr/74985/edbca/.conan/data/ruby/3.1.0/_/_/build/be0af55287e289adf24870a7f45a92f7dffcc5a5/build-release -rx86_64-darwin-fake  -I`cd /Users/jenkins/w/prod-v1/bsr/74985/edbca/.conan/data/ruby/3.1.0/_/_/build/be0af55287e289adf24870a7f45a92f7dffcc5a5/src/lib; pwd` --disable-gems -r./x86_64-darwin-fake /Users/jenkins/w/prod-v1/bsr/74985/edbca/.conan/data/ruby/3.1.0/_/_/build/be0af55287e289adf24870a7f45a92f7dffcc5a5/src/tool/rbinstall.rb --make="/Applications/conan/xcode/13.0/Xcode.app/Contents/Developer/usr/bin/make" --dest-dir="" --extout=".ext" --ext-build-dir="./ext" --mflags="- --jobserver-fds=4,5 -j" --make-flags=" --jobserver-fds=4,5 -j" --data-mode=0644 --prog-mode=0755 --installed-list .installed.list --mantype="doc"  --gnumake --install=local
installing binary commands:         //bin
configure: WARNING: using cross tools not prefixed with host triplet
/Users/jenkins/w/prod-v1/bsr/74985/edbca/.conan/data/ruby/3.1.0/_/_/build/be0af55287e289adf24870a7f45a92f7dffcc5a5/src/configure: line 12542: pkg-config: command not found
configure: WARNING: result yes guessed because of cross compilation
configure: WARNING: using cross tools not prefixed with host triplet
/Users/jenkins/w/prod-v1/bsr/74985/edbca/.conan/data/ruby/3.1.0/_/_/build/be0af55287e289adf24870a7f45a92f7dffcc5a5/src/configure: line 13651: pkg-config: command not found
configure: WARNING: result yes guessed because of cross compilation
ln -sf ../../../.ext/include/x86_64-darwin/rb_mjit_min_header-3.1.0.h include/ruby-3.1.0/x86_64-darwin/rb_mjit_min_header-3.1.0.h
/Applications/conan/xcode/13.0/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: file: openssl.a(openssl_missing.o) has no symbols
/Users/jenkins/w/prod-v1/bsr/74985/edbca/.conan/data/ruby/3.1.0/_/_/build/be0af55287e289adf24870a7f45a92f7dffcc5a5/src/lib/fileutils.rb:1395:in `initialize': Operation not permitted @ rb_sysopen - //bin/ruby (Errno::EPERM)
	from /Users/jenkins/w/prod-v1/bsr/74985/edbca/.conan/data/ruby/3.1.0/_/_/build/be0af55287e289adf24870a7f45a92f7dffcc5a5/src/lib/fileutils.rb:1395:in `open'
	from /Users/jenkins/w/prod-v1/bsr/74985/edbca/.conan/data/ruby/3.1.0/_/_/build/be0af55287e289adf24870a7f45a92f7dffcc5a5/src/lib/fileutils.rb:1395:in `block in copy_file'
	from /Users/jenkins/w/prod-v1/bsr/74985/edbca/.conan/data/ruby/3.1.0/_/_/build/be0af55287e289adf24870a7f45a92f7dffcc5a5/src/lib/fileutils.rb:1394:in `open'
	from /Users/jenkins/w/prod-v1/bsr/74985/edbca/.conan/data/ruby/3.1.0/_/_/build/be0af55287e289adf24870a7f45a92f7dffcc5a5/src/lib/fileutils.rb:1394:in `copy_file'
	from /Users/jenkins/w/prod-v1/bsr/74985/edbca/.conan/data/ruby/3.1.0/_/_/build/be0af55287e289adf24870a7f45a92f7dffcc5a5/src/lib/fileutils.rb:504:in `copy_file'
	from /Users/jenkins/w/prod-v1/bsr/74985/edbca/.conan/data/ruby/3.1.0/_/_/build/be0af55287e289adf24870a7f45a92f7dffcc5a5/src/lib/fileutils.rb:868:in `block in install'
	from /Users/jenkins/w/prod-v1/bsr/74985/edbca/.conan/data/ruby/3.1.0/_/_/build/be0af55287e289adf24870a7f45a92f7dffcc5a5/src/lib/fileutils.rb:1577:in `block in fu_each_src_dest'
	from /Users/jenkins/w/prod-v1/bsr/74985/edbca/.conan/data/ruby/3.1.0/_/_/build/be0af55287e289adf24870a7f45a92f7dffcc5a5/src/lib/fileutils.rb:1591:in `fu_each_src_dest0'
	from /Users/jenkins/w/prod-v1/bsr/74985/edbca/.conan/data/ruby/3.1.0/_/_/build/be0af55287e289adf24870a7f45a92f7dffcc5a5/src/lib/fileutils.rb:1575:in `fu_each_src_dest'
	from /Users/jenkins/w/prod-v1/bsr/74985/edbca/.conan/data/ruby/3.1.0/_/_/build/be0af55287e289adf24870a7f45a92f7dffcc5a5/src/lib/fileutils.rb:864:in `install'
	from /Users/jenkins/w/prod-v1/bsr/74985/edbca/.conan/data/ruby/3.1.0/_/_/build/be0af55287e289adf24870a7f45a92f7dffcc5a5/src/tool/rbinstall.rb:190:in `install'
	from /Users/jenkins/w/prod-v1/bsr/74985/edbca/.conan/data/ruby/3.1.0/_/_/build/be0af55287e289adf24870a7f45a92f7dffcc5a5/src/tool/rbinstall.rb:375:in `block in <main>'
	from /Users/jenkins/w/prod-v1/bsr/74985/edbca/.conan/data/ruby/3.1.0/_/_/build/be0af55287e289adf24870a7f45a92f7dffcc5a5/src/tool/rbinstall.rb:1127:in `block in <main>'
	from /Users/jenkins/w/prod-v1/bsr/74985/edbca/.conan/data/ruby/3.1.0/_/_/build/be0af55287e289adf24870a7f45a92f7dffcc5a5/src/tool/rbinstall.rb:1124:in `each'
	from /Users/jenkins/w/prod-v1/bsr/74985/edbca/.conan/data/ruby/3.1.0/_/_/build/be0af55287e289adf24870a7f45a92f7dffcc5a5/src/tool/rbinstall.rb:1124:in `<main>'
make: *** [do-install-local] Error 1
ERROR: ruby/3.1.0: Error in package() method, line 205
	autotools.make(target="install-local")
```

**Comment by jmarrec** [comment]
Seems like having ruby as both a requires and a tool_requires might be a problem, because when cross-compiling from x86_64 mac to armv8, one of the two archs is missing the binary and it therefore fails.

**Comment by jmarrec** [comment]
Seems that having ruby as both a `requires` and a `tool_requires` in the test_package is problematic because one of them has to be missing the prebuilt binary?

**Comment by jmarrec** [comment]
@valgur @samuel-emrys Could you please help me out with the testing that fails when cross building? https://github.com/conan-io/conan-center-index/pull/18338#issuecomment-1868224198

**Comment by jmarrec** [comment]
I just tried cross compiling on an apple silicon mac from armv8 to x86_64, which worked because I had built the armv8 (native) binary beforehand.

```
conan create --name ruby --version 3.1.0 --build missing -b build_type=Release -o '*/*:shared=False' -pr:h profile_osx_14_libcpp_17_apple-clang_release_64-shared-False -pr:b profile_osx_14_libcpp_17_apple-clang_release_armv8 

======== Input profiles ========
Profile host:
[settings]
arch=x86_64
build_type=Release
compiler=apple-clang
compiler.cppstd=17
compiler.libcxx=libc++
compiler.version=15
os=Macos
[options]
*/*:shared=False
[conf]
tools.system.package_manager:mode=install
tools.system.package_manager:sudo=True
tools.apple:sdk_path=/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.sdk

Profile build:
[settings]
arch=armv8
build_type=Release
compiler=apple-clang
compiler.cppstd=17
compiler.libcxx=libc++
compiler.version=15
os=Macos
[conf]
tools.system.package_manager:mode=install
tools.system.package_manager:sudo=True
tools.apple:sdk_path=/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.sdk
```

File: recipes/ruby/all/test_package/conanfile.py [hunk 9]
```diff
@@ -1,21 +1,71 @@
...
-    settings = "os", "compiler", "build_type", "arch"
-    generators = "CMakeDeps", "CMakeToolchain"
+    settings = "os", "arch", "compiler", "build_type"
+    generators = "CMakeDeps", "VirtualRunEnv"
+
+    def requirements(self):
+        self.requires(self.tested_reference_str)
+
+    def build_requirements(self):
+        self.tool_requires(self.tested_reference_str)
```
Comment (valgur) [inline recipes/ruby/all/test_package/conanfile.py:18]: You should be able to get rid of the "missing binary" error in CI by avoiding a `tool_requires()` on the tested package.
```diff
@@ -14,9 +14,6 @@
     def requirements(self):
         self.requires(self.tested_reference_str)
 
-    def build_requirements(self):
-        self.tool_requires(self.tested_reference_str)
-
     def layout(self):
         cmake_layout(self)
 
@@ -44,7 +41,7 @@
     def _test_ruby_executable_version(self):
         # test executable
         output = StringIO()
-        self.run("ruby --version", output, env="conanbuild")
+        self.run("ruby --version", output, env="conanrun")
         output_str = str(output.getvalue()).strip()
         self.output.info(f"Installed version: {output_str}")
         assert_ruby_version = f"ruby {self._ruby_version()}"
@@ -54,7 +51,7 @@
     def _test_ruby_execute(self):
         # test executable
         output = StringIO()
-        self.run('ruby -e "puts RUBY_VERSION"', output, env="conanbuild")
+        self.run('ruby -e "puts RUBY_VERSION"', output, env="conanrun")
         output_str = str(output.getvalue()).strip()
         self.output.info(f'ruby -e "puts RUBY_VERSION": "{output_str}"')
         assert_ruby_version = self._ruby_version()
```

Comment (jmarrec) [reply recipes/ruby/all/test_package/conanfile.py]: I had been advised to use a tool_requires. I've just tried removing it, let's see what CI says...
I did have to use `self.runenv_info.prepend_path("PATH", os.path.join(self.package_folder, "bin"))`

**Comment by jmarrec** [comment]
Ok, so we build miniruby, then we need to execute miniruby. Which is going to search for the dylibs.

The conanrunenv does have the `LD_LIBRARY_PATH` and `DYLD_LIBRARY_PATH`, but we don't have it activated. Not sure how to ensure these are defined when I build.

**Pull request closed without merging**